File: 8.回溯/9.分割回文串/131.分割回文串.py
#
# @lc app=leetcode.cn id=131 lang=python3
#
# [131] 分割回文串
#
from mytools import *
# @lc code=start
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        n = len(s)
        is_palin = [[False] * n for _ in range(n)]
        for i in range(n - 1, -1, -1):
            for j in range(i, n):
                if s[i] == s[j] and (j - i + 1 <= 3 or is_palin[i + 1][j - 1]):
                    is_palin[i][j] = True
        def dfs(sta, pack: List, res: List):
            if sta == n:
                res.append(pack.copy())
                return
            for eds in range(sta, n):
                sub = s[sta:eds + 1]
                if is_palin[sta][eds]:
                    pack.append(sub)
                    dfs(eds + 1, pack, res)
                    pack.pop()
        res = []
        dfs(0, [], res)
        return res

# @lc code=end

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def dfs(sta, pack: List, res: List):
            if sta == len(s):
                res.append(pack.copy())
                return
            for i in range(sta, len(s)):
                sub = s[sta:i + 1]
                if sub != sub[::-1]:
                    continue
                pack.append(sub)
                dfs(i + 1, pack, res)
                pack.pop()
        res = []
        dfs(0, [], res)
        return res

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def palin(s: str):
            l, r = 0, len(s) - 1
            while l <= r:
                if s[l] != s[r]: return False
                l += 1
                r -= 1
            return True
        
        def check(pack: List):
            for s in pack:
                if not palin(s): return False
            return True

        def dfs(sta, pack: List, res: List, _len):
            if _len == len(s):
                if check(pack): res.append(pack.copy())
                return
            for i in range(sta, len(s)):
                pack.append(s[sta:i + 1])
                _len += i + 1 - sta
                dfs(i + 1, pack, res, _len)
                _len -= i + 1 - sta
                pack.pop()
        res = []
        dfs(0, [], res, 0)
        return res

File: 8.回溯/9.分割回文串/131.分割回文串.cpp
/*
 * @lc app=leetcode.cn id=131 lang=cpp
 *
 * [131] 分割回文串
 */
#include "tools.h"
using namespace std;
// @lc code=start
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        vector<string> pack;
        dfs(s, 0, s.size(), res, pack);
        return res;
    }
    bool check(string& s)
    {
        int l = 0, r = s.size() - 1;
        while (l <= r)
        {
            if (s[l] != s[r]) return 0;
            l ++, r --;
        }
        return 1;
    }
    void dfs(string& s, int sta, int n, vector<vector<string>>& res, vector<string>& pack)
    {
        if (sta >= n)
        {
            res.emplace_back(pack);
            return;
        }
        for (int i = sta; i < n; i ++ )
        {
            string sub = s.substr(sta, i - sta + 1);
            if (! check(sub)) continue;
            pack.emplace_back(sub);
            dfs(s, i + 1, n, res, pack);
            pack.pop_back();
        }
    }
};
// @lc code=end

